typed-racket-lib/typed-racket/types/subtype.rkt

```diff
@@ -384,397 +396,560 @@
+               (subtype* A t1 b2)]
+              [(_ (PolyDots: vs2 b2))
+               #:when (and (null? (fv b2))
+                           (null? (fi b2)))
+               (subtype* A t1 b2)]
+              [(_ _) #f]))))])]))
+
+
+
+(define-switch (subtype-switch t1 t2 A continue)
+  ;; NOTE: keep these in alphabetical order
+  ;; or ease of finding cases
+  [(case: App _)
+   (let ([A (remember t1 t2 A)])
+     (with-updated-seen A
+       (let ([t1 (resolve-once t1)])
+         ;; check needed for if a name that hasn't been resolved yet
+         (and (Type? t1) (subtype* A t1 t2)))))]
+  [(case: Async-Channel (Async-Channel: elem1))
+   (match t2
+     [(? Async-ChannelTop?) A]
+     [(Async-Channel: elem2) (type-equiv? A elem1 elem2)]
+     [(Evt: evt-t) (subtype* A elem1 evt-t)]
+     [_ (continue)])]
+  [(case: Base (Base: kind _ pred numeric?))
+   (match t2
+     [(Sequence: (list seq-t))
+      (cond
+        [(assq kind `((FlVector . ,-Flonum)
+                      (ExtFlVector . ,-ExtFlonum)
+                      (FxVector . ,-Fixnum)
+                      (String . ,-Char)
+                      (Bytes . ,-Byte)
+                      (Input-Port . ,-Nat)))
+         => (λ (p) (subtype* A (cdr p) seq-t))]
+        [numeric?
+         (define type
+           ;; FIXME: thread the store through here
+           (for/or ([num-t (in-list (list -Byte -Index -NonNegFixnum -Nat))])
+             (or (and (subtype* A t1 num-t) num-t))))
+         (if type
+             (subtype* A type seq-t)
+             #f)]
+        [else #f])]
+     [(Evt: evt-t)
+      (cond
+        [(memq kind '(Semaphore
+                      Output-Port
+                      Input-Port
+                      TCP-Listener
+                      Thread
+                      Subprocess
+                      Will-Executor))
+         (subtype* A t1 evt-t)]
+        ;; FIXME: change Univ to Place-Message-Allowed if/when that type is defined
+        [(and (Univ? evt-t) (memq kind '(Place Base-Place-Channel)))
+         A]
+        [(eq? kind 'LogReceiver) (subtype* A
+                                           (make-HeterogeneousVector
+                                            (list -Symbol -String Univ
+                                                  (Un (-val #f) -Symbol)))
+                                           evt-t)]
+        [else #f])]
+     [_ (continue)])]
+  [(case: Box (Box: elem1))
+   (match t2
+     [(? BoxTop?) A]
+     [(Box: elem2) (type-equiv? A elem1 elem2)]
+     [_ (continue)])]
+  [(case: Channel (Channel: elem1))
+   (match t2
+     [(? ChannelTop?) A]
+     [(Channel: elem2) (type-equiv? A elem1 elem2)]
+     [(Evt: evt-t) (subtype* A elem1 evt-t)]
+     [_ (continue)])]
+  [(case: Class (Class: row inits fields methods augments init-rest))
+   (match t2
+     [(? ClassTop?) A]
+     [(Class: row* inits* fields* methods* augments* init-rest*)
+      ;; TODO: should the result be folded instead?
```

Comment:
It means the body uses `andmap` and does not thread `A` through the various recursive calls to `subtype*`, and the person who wrote this way back when was unsure if that was the correct choice.